tests/test_pandas.py

```diff
@@ -133,17 +135,74 @@ def test_isna() -> None:
+    # There is a proposal being floated for a StrictTypeGuard that will have more rigid narrowing semantics
+    # In the test cases below, a commented out assertion will be included to document the optimal test result
+    nullable1 = random.choice(["value", None, pd.NA, pd.NaT])
+    if pd.notna(nullable1):
+        check(assert_type(nullable1, str), str)
+    if not pd.isna(nullable1):
+        # TODO: This is the true type (see comments above on the limitations of TypeGuard)
+        # check(assert_type(nullable1, str), str)
+        check(assert_type(nullable1, Union[str, NaTType, NAType, None]), str)
+    if pd.isna(nullable1):
+        assert_type(nullable1, Union[NaTType, NAType, None])
+    if not pd.notna(nullable1):
+        # TODO: This is the true type (see comments above on the limitations of TypeGuard)
+        # assert_type(nullable1, Union[NaTType, NAType, None])
+        assert_type(nullable1, Union[str, NaTType, NAType, None])
+
+    nullable2 = random.choice([2, None])
+    if pd.notna(nullable2):
+        check(assert_type(nullable2, int), int)
+    if not pd.isna(nullable2):
+        # TODO: This is the true type (see comments above on the limitations of TypeGuard)
+        # check(assert_type(nullable2, int), int)
+        check(assert_type(nullable2, Union[int, None]), int)
+    if pd.isna(nullable2):
+        # TODO: This is the true type (see comments above on the limitations of TypeGuard)
+        # check(assert_type(nullable2, None), type(None))
+        check(assert_type(nullable2, Union[NaTType, NAType, None]), type(None))
+    if not pd.notna(nullable2):
+        # TODO: This is the true type (see comments above on the limitations of TypeGuard)
+        # check(assert_type(nullable2, None), type(None))
+        assert_type(nullable2, Union[int, None])  # TODO: MyPy result
+        assert_type(
+            nullable2, Union[int, NaTType, NAType, None]
+        )  # TODO: Pyright result
+
+    nullable3 = random.choice([True, None, pd.NA])
+    if pd.notna(nullable3):
+        check(assert_type(nullable3, bool), bool)
+    if not pd.isna(nullable3):
+        # TODO: This is the true type (see comments above on the limitations of TypeGuard)
+        # check(assert_type(nullable3, bool), bool)
+        check(assert_type(nullable3, Union[bool, NAType, None]), bool)
+    if pd.isna(nullable3):
+        # TODO: This is the true type (see comments above on the limitations of TypeGuard)
+        # assert_type(nullable3, Union[NAType, None])
+        assert_type(nullable3, Union[NaTType, NAType, None])
+    if not pd.notna(nullable3):
+        # TODO: This is the true type (see comments above on the limitations of TypeGuard)
+        # assert_type(nullable3, Union[NAType, None])
+        assert_type(nullable3, Union[bool, NAType, None])  # TODO: Mypy result
+        assert_type(
+            nullable3, Union[bool, NaTType, NAType, None]
+        )  # TODO: Pyright result
```

Comment:
I added a lot more test cases and tested the negative condition as well. I also documented what the correct type should be - which isn't always supported by the way `TypeGuard` currently works (see comments on `StrictTypeGuard`.

However, this uncovered conflicting behavior in Mypy and Pyright. My 2 cents on this, is that I believe Mypy is incorrect, and Pyright is correct. This is just an extension on how TypeGuard generally behaves - a simple type narrowing from A to B, with no regard for unioned types and set arithmetic. And as long as that's true, then Mypy shouldn't be narrowing beyond that, as it does on line 203 and 183 for example.